docs/csharp/tutorials/pattern-matching.md

```diff
@@ -0,0 +1,325 @@
+
+[!code-csharp[ExternalSystems](../../../samples/csharp/tutorials/patterns/start/toll-calculator/ExternalSystems.cs)]
+
+You can download the starter code from the [dotnet/samples](https://github.com/dotnet/samples/tree/master/csharp/tutorials/patterns/start) GitHub repository. You can see that the vehicle classes are from different systems, and are in different namespaces. No common base class, other than `System.Object` can be leveraged.
+
+## Pattern matching designs
+
+The scenario used in this tutorial highlights the kinds of problems that are well suited to use pattern matching to solve: 
+
+- The objects you need to work with aren't in an object hierarchy that matches your goals. You may be working with classes that are part of unrelated systems.
+- The functionality you're adding isn't part of the core abstraction for these classes. The toll paid by a vehicle *changes* for different types of vehicles, but the toll isn't a core function of the vehicle.
+
+The pattern matching features in C# make it easier to build software when the *shape* of the data and the *operations* on that data are not described together.
+
+## Implement the basic toll calculations
+
+The most basic toll calculation relies only on the vehicle type:
+
+- A `Car` is $2.00.
+- A `Taxi` is $3.50.
+- A `Bus` is $5.00.
+- A `DeliveryTruck` is $10.00
+
+Create a new `TollCalculator` class, and implement the You can implement that using this version of a `TollCalculator` class:
+
+```csharp
+using System;
+using CommercialRegistration;
+using ConsumerVehicleRegistration;
+using LiveryRegistration;
+
+namespace toll_calculator
+{
+    public class TollCalculator
+    {
+        public decimal CalculateToll(object vehicle) =>
+            vehicle switch
+        {
+            Car c => 2.00m,
+            Taxi t => 3.50m,
+            Bus b => 5.00m,
+            DeliveryTruck t => 10.00m,
+            { } => throw new ArgumentException(message: "Not a known vehicle type", paramName: nameof(vehicle)),
+            null => throw new ArgumentNullException(nameof(vehicle))
+        };
+    }
+}
+```
+
+The preceding code uses a **switch expression** that tests the **type pattern**. A **switch expression** begins with the variable, `vehicle` in the preceding code, followed by the `switch` keyword. Next comes all the switch arms inside curly braces. The `switch` expression makes other refinements to the ceremony that surrounded the `switch` statement. The `case` keyword is omitted, and the result of each arm is an expression. The last two arms show a new language feature. The `{ }` case matches any non-null object that didn't match an earlier arm. This arm catches any incorrect types passed to this method. Finally, the `null` pattern catches when `null` is passed to this method. The `null` pattern can be last because the other type patterns match only a non-null object of the correct type.
```

Comment:
I think it's OK to point out this is not the same as a switch **statement**. If someone is new to this, they may not realize they are two different things. 

"A **switch expression** (not the same as a switch **statement**) begins with the variable..." 

Maybe a link in this first use of **switch expression** and **switch statement** to the keyword docs?

OOOOOR maybe I'm totally wrong in that thought assumption and they really are the same just expressed differently? With the rule difference of not having `case` or `default` it makes it feel like it's a different concept.

----

Also, the use of the word **ceremony** feels odd... "syntax" ??

Reply:
I did add the clarifying remarks.

We haven't written the reference docs for the switch expression yet, so I only added the link to the switch statement. 😭